### PR TITLE
Refactor `Study.trials_dataframe`

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

See https://github.com/optuna/optuna/pull/1904. Basically the same, but for `Study.trials_dataframe` instead of `Study.optimize`. 

I could see reasons for not merging this since it adds another level of indirection and a new file partially to slim down `study.py`. However, we can isolate the dependency of `pandas` to a more confined scope, namely `_dataframe.py`, rather than `study.py`. Additionally, the logic is suited as a free function since it didn't used to rely on `self` besides from a single public API call, to obtain a list of all trials.

## Description of the changes

See above.

### CODE REVIEW

File: optuna/_dataframe.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+with try_import() as _imports:
+    # `trials_dataframe` is disabled if pandas is not available.
+    import pandas as pd  # NOQA
+
+# Required for type annotation in `Study.trials_dataframe`.
+if not _imports.is_successful():
+    pd = object  # type: ignore # NOQA
+
+
+def _trials_dataframe(
```
Comment (hvy) [inline optuna/_dataframe.py:22]: The implementation of this function is just a copy from `Study.trials_dataframe`.

File: optuna/_dataframe.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from typing import Set
+from typing import Tuple
+
+import optuna
+from optuna._imports import try_import
+from optuna.trial._state import TrialState
+
+
+with try_import() as _imports:
+    # `trials_dataframe` is disabled if pandas is not available.
```
Comment (toshihikoyanase) [inline optuna/_dataframe.py:14]: A nit, but `trials_dataframe` seems to be either `Study.trials_dataframe` or `_trials_dataframe`.
```suggestion
    # `Study.trials_dataframe` is disabled if pandas is not available.
```

File: optuna/_dataframe.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# Required for type annotation in `Study.trials_dataframe`.
+if not _imports.is_successful():
+    pd = object  # type: ignore # NOQA
+
+
+def _trials_dataframe(
+    study: "optuna.Study",
+    attrs: Tuple[str, ...],
+    multi_index: bool,
+) -> "pd.DataFrame":
```
Comment (toshihikoyanase) [inline optuna/_dataframe.py:26]: Do we need the trailing comma? The number of arguments is not so large, and I'm not sure if it is needed here.
```suggestion
    study: "optuna.Study", attrs: Tuple[str, ...], multi_index: bool
) -> "pd.DataFrame":
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. It basically looks good to me, but I have some small comments.

File: optuna/_dataframe.py (hunk 3 above)
Comment (hvy) [reply optuna/_dataframe.py:26]: You're right, let's remove the comma.

**Comment by hvy** [comment]
Thanks for the review! Comments should be addressed now, PTAL.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your swift action. LGTM!

File: optuna/_dataframe.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from typing import Tuple
+
+import optuna
+from optuna._imports import try_import
+from optuna.trial._state import TrialState
+
+
+with try_import() as _imports:
+    # `Study.trials_dataframe` is disabled if pandas is not available.
+    import pandas as pd  # NOQA
```
Comment (himkt) [inline optuna/_dataframe.py:15]: [question]

Is `# NOQA` still needed here? (I'm wondering if we can remove it, as there is no variable assignment here.)

**Review by himkt — APPROVED** [review]
LGTM with a nits question.

File: optuna/_dataframe.py (hunk 4 above)
Comment (hvy) [reply optuna/_dataframe.py:15]: I think you're right. Removed it.

**Comment by hvy** [comment]
Removed the `#NOQA` as pointed out in review, PTAL.

**Pull request merged**